dbms/src/Common/BitpackingPrimitives.h

```diff
@@ -0,0 +1,340 @@
+        if (bitwidth < sizeof(T) * 8 && bitwidth != 0)
+        {
+            if constexpr (is_signed)
+            {
+                RUNTIME_ASSERT(max_value <= (T(1) << (bitwidth - 1)) - 1);
+                RUNTIME_ASSERT(min_value >= (T(-1) * ((T(1) << (bitwidth - 1)) - 1) - 1));
+            }
+            else
+            {
+                RUNTIME_ASSERT(max_value <= (T(1) << (bitwidth)) - 1);
+            }
+        }
+        if constexpr (round_to_next_byte)
+        {
+            return (bitwidth / 8 + (bitwidth % 8 != 0)) * 8;
+        }
+        return bitwidth;
+    }
+
+    // Sign bit extension
+    template <typename T, typename TU = typename std::make_unsigned<T>::type>
+    static void signExtend(unsigned char * dst, UInt8 width)
+    {
+        T const mask = static_cast<T>(TU(1) << (width - 1));
+        for (size_t i = 0; i < BitpackingPrimitives::BITPACKING_ALGORITHM_GROUP_SIZE; ++i)
+        {
+            T value = unalignedLoad<T>(dst + i * sizeof(T));
+            value = static_cast<T>(value & ((TU(1) << width) - TU(1)));
+            T result = (value ^ mask) - mask;
+            unalignedStore<T>(dst + i * sizeof(T), result);
+        }
+    }
+
+    // Prevent compression at widths that are ineffective
+    template <typename T>
+    constexpr static UInt8 getEffectiveWidth(UInt8 width)
+    {
+        UInt8 bits_of_type = sizeof(T) * 8;
+        UInt8 type_size = sizeof(T);
+        if (width + type_size > bits_of_type)
+        {
+            return bits_of_type;
+        }
+        return width;
+    }
+
+    template <typename T>
+    static void packGroup(unsigned char * dst, const T * values, UInt8 width)
+    {
+        if constexpr (std::is_same<T, Int8>::value || std::is_same<T, UInt8>::value)
+        {
+            fastpforlib::fastpack(
+                reinterpret_cast<const UInt8 *>(values),
+                reinterpret_cast<UInt8 *>(dst),
+                static_cast<UInt32>(width));
+        }
+        else if constexpr (std::is_same<T, Int16>::value || std::is_same<T, UInt16>::value)
+        {
+            fastpforlib::fastpack(
+                reinterpret_cast<const UInt16 *>(values),
+                reinterpret_cast<UInt16 *>(dst),
+                static_cast<UInt32>(width));
+        }
+        else if constexpr (std::is_same<T, Int32>::value || std::is_same<T, UInt32>::value)
+        {
+            fastpforlib::fastpack(
+                reinterpret_cast<const UInt32 *>(values),
+                reinterpret_cast<UInt32 *>(dst),
+                static_cast<UInt32>(width));
+        }
+        else if constexpr (std::is_same<T, Int64>::value || std::is_same<T, UInt64>::value)
+        {
+            fastpforlib::fastpack(
+                reinterpret_cast<const UInt64 *>(values),
+                reinterpret_cast<UInt32 *>(dst),
+                static_cast<UInt32>(width));
+        }
+        else
+        {
+            throw Exception("Unsupported type for bitpacking");
```

Comment:
I think it can actually be made into a compiler error?

Reply:
```cpp
            static_assert(false, "Unsupported type for bitpacking");
```

Reply:
revert, keep `throw Exception` since clang13 does not support `static_assert(false)`